Add support for PayPal web experience profiles.

jimabramson opened the pull request:
XCOM-319

@rlucioni @clintonb FYI @Nickersoft 

I still need to add tests, but would like to start getting feedback on the overall approach.  PR consists of:

* Management command:
  * wrappers around PayPal API for list/create/show/update/delete.  Initially these seemed important as manually testing directly via the REST API is cumbersome.   However, most aren't actually part of any deployment workflow.  Since there's a substantial amount of code around this, I think it should have tests, but alternatively, we can just gut it and keep this a lot smaller.
  * `generate`, `enable`, and `disable` commands.  These, respectively: 1) generate and store with PayPal a web experience profile based on values in configuration; 2) save the profile in a django model that we use when setting up a PayPal payment; 3) delete the profile from that model.
* Add settings to store the brand name and logo image under a named profile key.
* Add the PayPalWebProfile model and migration.   Very simple.

There will be a configuration PR shortly that corresponds to this change.

clintonb commented on ecommerce/extensions/payment/management/commands/paypal_profile.py:
```diff
@@ -0,0 +1,202 @@
+        action = None
+        try:
+            action = args.pop(0)
+            handler = getattr(self, 'handle_{}'.format(action))
```
A help message for these actions would be...helpful. :)

clintonb commented on ecommerce/extensions/payment/processors/paypal.py:
```diff
@@ -31,6 +32,7 @@ class Paypal(BasePaymentProcessor):
     """
 
     NAME = u'paypal'
+    WEB_PROFILE_NAME = 'openedx_default'
```
Should this be a setting? There doesn't seem to be a way to override it.

jimabramson replied:
There isn't.  At this stage I think there is no need for multiple profiles and so a hardcoded name would be fine.  It's partially a hedge against the fact that we may want to dynamically generate these in the future, or just pre-configure a limited amount of them - I wanted to avoid committing to either direction.

clintonb replied:
Okay.

rlucioni commented on ecommerce/extensions/payment/models.py:
```diff
@@ -25,5 +25,11 @@ class Meta(object):
     card_type = models.CharField(max_length=255, choices=CARD_TYPE_CHOICES, null=True, blank=True)
 
+
+class PayPalWebProfile(models.Model):
```
Minor: for consistency with our `Paypal` payment processor class, I'd name this `PaypalWebProfile`.

jimabramson commented:
updated and ready for re-review.

rlucioni commented on ecommerce/extensions/payment/management/commands/paypal_profile.py:
```diff
@@ -0,0 +1,229 @@
+        update [id] [json]  Update an existing profile using the specified JSON string.
+        delete [id]         Delete an existing profile. (Use -d to automatically disable when deleting.)
+        enable [id]         Enable the web profile in this Django application (send it in PayPal api calls).
+        disable [id]        Disable the web profile in this Django application (don't send in PayPal api calls).
```
Minor, but: s/api/API

rlucioni commented:
:+1:

jimabramson commented:
@rlucioni thanks, fixed caps with 'api'
@clintonb when you have time, lmk if the amended help string in the command module seems sufficient, or if you have additional concerns.

clintonb commented:
:+1:

jimabramson commented:
@clintonb pls review.

added: PaypalProfileAdminView
removed: settings related to profiles and the `generate` command intended to automate this stuff.  (see the last fixup commit to isolate what's changed since your last thumbsup)

jimabramson commented:
last commit fixed the quality issue

clintonb commented on ecommerce/extensions/payment/views.py:
```diff
@@ -216,3 +219,59 @@ def get(self, request):
             return redirect(receipt_url)
+
+
+class PaypalProfileAdminView(View):
```
It would be great if this were a proper REST API, but I won't push the matter.

jimabramson replied:
thanks :-)

it's already somewhat surreal that this exists at all, but following the example of the course migration, doing everything with GET really heightens the convenience factor.

clintonb commented on ecommerce/extensions/payment/views.py:
```diff
@@ -216,3 +219,59 @@ def get(self, request):
+
+class PaypalProfileAdminView(View):
+
+    ACTIONS = ('list', 'create', 'show', 'update', 'delete')
```
What about enable/disable?

jimabramson replied:
those can be done via django admin, so it seemed redundant to expose them here as well.

clintonb replied:
My concern here is that we will still be reliant on DevOps by not exposing
the enable/disable actions.

-- 
Clinton Blackburn

jimabramson replied:
OK, updated.

clintonb commented on ecommerce/extensions/payment/views.py:
```diff
@@ -216,3 +219,59 @@ def get(self, request):
+        err = StringIO()
+        log = StringIO()
+
+        root_logger = logging.getLogger()
```
Use the existing global, `logger`.

jimabramson replied:
fixed.

clintonb commented:
:+1: (after squashing)
